inc/Engine/Preload/Controller/PreloadUrl.php

```diff
@@ -58,27 +58,80 @@ public function __construct( Options_Data $options, Queue $queue, Cache $rocket_
 	 * @return void
 	 */
 	public function preload_url( string $url ) {
-		if ( $this->is_already_cached( $url ) ) {
+
+		$is_mobile = $this->options->get( 'do_caching_mobile_files', false );
+		if ( $this->is_already_cached( $url ) && ( ! $is_mobile || $this->is_already_cached( $url, true ) ) ) {
 			$this->query->make_status_complete( $url );
 			return;
 		}
-		wp_safe_remote_get(
-			user_trailingslashit( $url ),
+
+		$requests = [
 			[
-				'blocking' => false,
-				'timeout'  => 0.01,
-			]
-			);
-		if ( $this->options->get( 'cache_mobile', false ) ) {
-			wp_safe_remote_get(
-				user_trailingslashit( $url ),
-				[
+				'url'       => $url,
+				'is_mobile' => false,
+				'headers'   => [
 					'blocking'   => false,
 					'timeout'    => 0.01,
+					'user-agent' => 'WP Rocket/Preload',
+				],
+			],
+		];
+
+		if ( $is_mobile ) {
+			$requests[] = [
+				'url'       => $url,
+				'headers'   => [
 					'user-agent' => $this->get_mobile_user_agent_prefix(),
-				]
+				],
+				'is_mobile' => true,
+			];
+		}
+
+		/**
+		 * Filters to modify requests done to preload an url.
+		 *
+		 * @param array $requests Requests that will be done.
+		 */
+		$requests = apply_filters( 'rocket_preload_before_preload_url', $requests );
+
+		$requests = array_filter( $requests );
+
+		foreach ( $requests as $request ) {
+			if ( isset( $request['url'] ) && is_string( $request['url'] ) ) {
```

Comment:
could change the conditional into an early bail-out if url is not set or not a string

Reply:
Can't you do the following?

```
if ( ! isset( $request['url'] ) ||  ! is_string( $request['url'] ) ) {
    continue;
}
```